codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/customization/UploadPartPresignedURLMiddleware.kt

```diff
@@ -0,0 +1,45 @@
+package software.amazon.smithy.aws.swift.codegen.customization
+
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.swift.codegen.Middleware
+import software.amazon.smithy.swift.codegen.SwiftWriter
+import software.amazon.smithy.swift.codegen.swiftmodules.SmithyTypes
+
+// This middleware is intended only for use with S3 `UploadPart`, and only for use when
+// creating a presigned URL.
+//
+// Generates a middleware that writes S3 object metadata into the HTTP query string.
+class UploadPartPresignedURLMiddleware(
+    val inputSymbol: Symbol,
+    outputSymbol: Symbol,
+    outputErrorSymbol: Symbol,
+    private val writer: SwiftWriter
+) : Middleware(writer, inputSymbol) {
+    override val typeName = "UploadPartPresignedURLMiddleware"
+
+    override fun generateInit() {
+        writer.write("public init() {}")
+    }
+
+    override fun renderExtensions() {
+        writer.write(
+            """
+            extension $typeName: Smithy.RequestMessageSerializer {
+                public typealias InputType = ${inputSymbol.name}
+                public typealias RequestType = SmithyHTTPAPI.HTTPRequest
+                
+                public func apply(input: InputType, builder: SmithyHTTPAPI.HTTPRequestBuilder, attributes: Smithy.Context) throws {
+                    if let partNumber = input.partNumber {
+                        let queryItem = Smithy.URIQueryItem(name: "partNumber".urlPercentEncoding(), value: Swift.String(partNumber).urlPercentEncoding())
+                        builder.withQueryItem(queryItem)
+                    }
+                    if let uploadId = input.uploadId {
+                        let queryItem = Smithy.URIQueryItem(name: "uploadId".urlPercentEncoding(), value: Swift.String(uploadId).urlPercentEncoding())
+                        builder.withQueryItem(queryItem)
+                    }
```

Comment:
Wondering if there is a way to generate this.

Reply:
Query item provider is already generated [here](https://github.com/awslabs/aws-sdk-swift/blob/ba7ed92d600ea3068fcc9a8b572dcb8e41d2df4d/Sources/Services/AWSS3/Sources/AWSS3/Models.swift#L16398-L16417). It handles adding input values with [@httpQuery trait ](https://smithy.io/2.0/spec/http-bindings.html#httpquery-trait) in Smithy model for the service to the request at runtime. 

The `UploadPartInput.queryItemProvider` gets used [here](https://github.com/awslabs/aws-sdk-swift/blob/ba7ed92d600ea3068fcc9a8b572dcb8e41d2df4d/Sources/Services/AWSS3/Sources/AWSS3/S3Client.swift#L7923) for normal request flow, and to add `QueryItemMiddleware` for the `presignURL` method, only change that needs to be made is probably adding the middleware codegen back for `UploadPart` in [here](https://github.com/awslabs/aws-sdk-swift/blob/ba7ed92d600ea3068fcc9a8b572dcb8e41d2df4d/codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/customization/presignable/PresignableUrlIntegration.kt#L198-L207). I.e., this custom middleware is not necessary.